File: packages/flutter_blue_plus_android/android/src/main/java/com/lib/flutter_blue_plus/model/GlucoseMeasurementRecord.java
package com.lib.flutter_blue_plus.model;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.GregorianCalendar;
import java.util.Locale;

public class GlucoseMeasurementRecord implements Parcelable {
    private int sequenceNumber;
    private GregorianCalendar calendar;
    private int timeOffset;
    private GlucoseConcentrationMeasurementUnit glucoseConcentrationMeasurementUnit;
    private float glucoseConcentrationValue;
    private int type;
    private int sampleLocationInteger;
    private String testBloodType;
    private String sampleLocation;
    private SensorStatusAnnunciation sensorStatusAnnunciation;

    public GlucoseMeasurementRecord() {
        this.sequenceNumber = 0;  // Số thứ tự của lần đo.
        this.calendar = new GregorianCalendar(Locale.UK); // Ngày giờ của lần đo, mặc định là múi giờ UK.
        this.timeOffset = 0; // Độ lệch thời gian so với thời điểm đo.
        this.glucoseConcentrationMeasurementUnit = GlucoseConcentrationMeasurementUnit.MOLES_PER_LITRE; // Đơn vị đo đường huyết (mol/L hoặc kg/L).
        this.glucoseConcentrationValue = 0f; // Giá trị đo nồng độ glucose.
        this.type = 0;  // Loại mẫu máu (Ví dụ: máu mao mạch, máu động mạch, huyết tương...).
        this.sampleLocationInteger = 0;  // Vị trí lấy mẫu (Ví dụ: ngón tay, dái tai...).
        this.testBloodType = "Capillary Whole blood";  // Chuỗi chứa loại máu đo (cập nhật dựa vào type).
        this.sampleLocation = "Earlobe"; // Chuỗi chứa vị trí lấy mẫu (cập nhật dựa vào sampleLocationInteger).
        this.sensorStatusAnnunciation = null; // Trạng thái của cảm biến đo (có thể null nếu không có dữ liệu).
        initializeBloodTypeAndLocation();
    }

    private void initializeBloodTypeAndLocation() {
        switch (type) {
            case 0:
                testBloodType = "Reserved for future use"; // Dùng cho mục đích sau này
                break;
            case 1:
                testBloodType = "Capillary Whole blood"; // Máu toàn phần mao mạch
                break;
            case 2:
                testBloodType = "Capillary Plasma"; // Huyết tương mao mạch
                break;
            case 3:
                testBloodType = "Venous Whole blood"; // Máu toàn phần tĩnh mạch
                break;
            case 4:
                testBloodType = "Venous Plasma"; // Huyết tương tĩnh mạch
                break;
            case 5:
                testBloodType = "Arterial Whole blood"; // Máu toàn phần động mạch
                break;
            case 6:
                testBloodType = "Arterial Plasma"; // Huyết tương động mạch
                break;
            case 7:
                testBloodType = "Undetermined Whole blood"; // Máu toàn phần chưa xác định
                break;
            case 8:
                testBloodType = "Undetermined Plasma"; // Huyết tương chưa xác định
                break;
            case 9:
                testBloodType = "Interstitial Fluid (ISF)"; // Dịch kẽ (ISF)
                break;
            case 10:
                testBloodType = "Control Solution"; // Dung dịch đối trứng
                break;
            default:
                testBloodType = "Reserved for future use"; // Dùng cho mục đích sau này
                break;
        }

        switch (sampleLocationInteger) {
            case 0:
                sampleLocation = "Reserved for future use"; // Dùng cho mục đích sau này
                break;
            case 1:
                sampleLocation = "Finger"; // Ngón tay
                break;
            case 2:
                sampleLocation = "Alternate Site Test (AST)"; // Xét nghiệm vị trí thay thế
                break;
            case 3:
                sampleLocation = "Earlobe"; // Dái tai
                break;
            case 4:
                sampleLocation = "Control solution"; // Dung dịch kiểm soát
                break;
            case 15:
                sampleLocation = "Sample Location value not available"; // Giá trị vị trí mẫu không khả dụng
                break;
            default:
                sampleLocation = "Reserved for future use"; // Dùng cho mục đích sau này
                break;
        }
    }

    public String convertGlucoseConcentrationValueToMilligramsPerDeciliter() {
        return (glucoseConcentrationValue * 100_000) + "mg/dL";
    }

    // CHUYỂN ĐỔI GIÁ TRỊ
    public enum GlucoseConcentrationMeasurementUnit {
        MOLES_PER_LITRE("mol/L"), KILOGRAM_PER_LITRE("kg/L");

        private final String unit;

        GlucoseConcentrationMeasurementUnit(String unit) {
            this.unit = unit;
        }
    }

    protected GlucoseMeasurementRecord(Parcel in) {
        sequenceNumber = in.readInt();
        timeOffset = in.readInt();
        glucoseConcentrationValue = in.readFloat();
        type = in.readInt();
        sampleLocationInteger = in.readInt();
        testBloodType = in.readString();
        sampleLocation = in.readString();
        glucoseConcentrationMeasurementUnit = (GlucoseConcentrationMeasurementUnit) in.readSerializable();
    }

    public static final Creator<GlucoseMeasurementRecord> CREATOR = new Creator<GlucoseMeasurementRecord>() {
        @Override
        public GlucoseMeasurementRecord createFromParcel(Parcel in) {
            return new GlucoseMeasurementRecord(in);
        }

        @Override
        public GlucoseMeasurementRecord[] newArray(int size) {
            return new GlucoseMeasurementRecord[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(sequenceNumber);
        dest.writeInt(timeOffset);
        dest.writeFloat(glucoseConcentrationValue);
        dest.writeInt(type);
        dest.writeInt(sampleLocationInteger);
        dest.writeString(testBloodType);
        dest.writeString(sampleLocation);
        dest.writeSerializable(glucoseConcentrationMeasurementUnit);
    }
}

File: packages/flutter_blue_plus_darwin/darwin/flutter_blue_plus_darwin/Sources/flutter_blue_plus_darwin/include/model/GlucoseMeasurementRecord.m
// GlucoseMeasurementRecord.m
#import "GlucoseMeasurementRecord.h"

@implementation GlucoseMeasurementRecord

// Phương thức khởi tạo
- (instancetype)initWithSequenceNumber:(NSInteger)sequenceNumber
                              calendar:(NSDate *)calendar
                            timeOffset:(NSInteger)timeOffset
   glucoseConcentrationMeasurementUnit:(GlucoseConcentrationMeasurementUnit)unit
             glucoseConcentrationValue:(float)value
                                  type:(NSInteger)type
                 sampleLocationInteger:(NSInteger)sampleLocationInteger
              sensorStatusAnnunciation:(nullable NSNumber *)status {
    self = [super init];
    if (self) {
        _sequenceNumber = sequenceNumber;
        _calendar = calendar;
        _timeOffset = timeOffset;
        _glucoseConcentrationMeasurementUnit = unit;
        _glucoseConcentrationValue = value;
        _type = type;
        _sampleLocationInteger = sampleLocationInteger;
        _sensorStatusAnnunciation = status;

        // Xử lý kiểu máu và vị trí mẫu
        [self updateTestBloodTypeAndSampleLocation];
    }
    return self;
}

// Phương thức chuyển đổi nồng độ glucose sang mg/dL
- (NSString *)convertGlucoseConcentrationValueToMilligramsPerDeciliter {
    return [NSString stringWithFormat:@"%.0f mg/dL", self.glucoseConcentrationValue * 100000];

}

// Cập nhật loại máu và vị trí mẫu
- (void)updateTestBloodTypeAndSampleLocation {
    switch (self.type) {
        case 0:
            _testBloodType = @"Reserved for future use";
            break;
        case 1:
            _testBloodType = @"Capillary Whole blood";
            break;
        case 2:
            _testBloodType = @"Capillary Plasma";
            break;
        case 3:
            _testBloodType = @"Venous Whole blood";
            break;
        case 4:
            _testBloodType = @"Venous Plasma";
            break;
        case 5:
            _testBloodType = @"Arterial Whole blood";
            break;
        case 6:
            _testBloodType = @"Arterial Plasma";
            break;
        case 7:
            _testBloodType = @"Undetermined Whole blood";
            break;
        case 8:
            _testBloodType = @"Undetermined Plasma";
            break;
        case 9:
            _testBloodType = @"Interstitial Fluid (ISF)";
            break;
        case 10:
            _testBloodType = @"Control Solution";
            break;
        default:
            _testBloodType = @"Reserved for future use";
            break;
    }

    switch (self.sampleLocationInteger) {
        case 0:
            _sampleLocation = @"Reserved for future use";
            break;
        case 1:
            _sampleLocation = @"Finger";
            break;
        case 2:
            _sampleLocation = @"Alternate Site Test (AST)";
            break;
        case 3:
            _sampleLocation = @"Earlobe";
            break;
        case 4:
            _sampleLocation = @"Control solution";
            break;
        case 15:
            _sampleLocation = @"Sample Location value not available";
            break;
        default:
            _sampleLocation = @"Reserved for future use";
            break;
    }
}

// Phương thức mã hóa đối tượng để truyền qua các activity khác
- (void)encodeWithCoder:(NSCoder *)coder {
    [coder encodeInteger:self.sequenceNumber forKey:@"sequenceNumber"];
    [coder encodeObject:self.calendar forKey:@"calendar"];
    [coder encodeInteger:self.timeOffset forKey:@"timeOffset"];
    [coder encodeInteger:self.glucoseConcentrationMeasurementUnit forKey:@"glucoseConcentrationMeasurementUnit"];
    [coder encodeFloat:self.glucoseConcentrationValue forKey:@"glucoseConcentrationValue"];
    [coder encodeInteger:self.type forKey:@"type"];
    [coder encodeInteger:self.sampleLocationInteger forKey:@"sampleLocationInteger"];
    [coder encodeObject:self.testBloodType forKey:@"testBloodType"];
    [coder encodeObject:self.sampleLocation forKey:@"sampleLocation"];
    [coder encodeObject:self.sensorStatusAnnunciation forKey:@"sensorStatusAnnunciation"];
}

// Phương thức giải mã đối tượng
- (nullable instancetype)initWithCoder:(NSCoder *)coder {
    NSInteger sequenceNumber = [coder decodeIntegerForKey:@"sequenceNumber"];
    NSDate *calendar = [coder decodeObjectForKey:@"calendar"];
    NSInteger timeOffset = [coder decodeIntegerForKey:@"timeOffset"];
    GlucoseConcentrationMeasurementUnit unit = [coder decodeIntegerForKey:@"glucoseConcentrationMeasurementUnit"];
    float glucoseConcentrationValue = [coder decodeFloatForKey:@"glucoseConcentrationValue"];
    NSInteger type = [coder decodeIntegerForKey:@"type"];
    NSInteger sampleLocationInteger = [coder decodeIntegerForKey:@"sampleLocationInteger"];
    NSString *testBloodType = [coder decodeObjectForKey:@"testBloodType"];
    NSString *sampleLocation = [coder decodeObjectForKey:@"sampleLocation"];
    NSNumber *sensorStatusAnnunciation = [coder decodeObjectForKey:@"sensorStatusAnnunciation"];

    self = [self initWithSequenceNumber:sequenceNumber
                               calendar:calendar
                             timeOffset:timeOffset
    glucoseConcentrationMeasurementUnit:unit
              glucoseConcentrationValue:glucoseConcentrationValue
                                   type:type
                  sampleLocationInteger:sampleLocationInteger
               sensorStatusAnnunciation:sensorStatusAnnunciation];

    return self;
}

@end
